Feature/Scrum 117 아티스트 선택 UI 구현

inwoodev opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경
- 기획에 따라 유저가 홈화면에 처음으로 접근시 자기가 선호하는 아티스트를 선택할 수 있는 화면이 있어야 합니다.

## 목표
- 유저가 해당 화면을 닫을 수 있게 화면 좌측 상단에 닫기 버튼을 구현합니다.
- 유저가 아티스트를 검색할 수 있는 searchBar를 구현합니다.
- 아티스트 목록을 볼 수 있는 collectionview를 구현합니다.
- 아티스트 선택 후 다음화면으로 이동할 수 있는 버튼을 하단에 구현합니다.

## 결과 (Optional)
![Simulator Screen Recording - iPhone 15 - 2024-06-19 at 02 12 15](https://github.com/project-lyrics/app-iOS/assets/69072471/00b42101-6892-4907-a9ef-05eddd6af304)

inwoodev committed:
refactor, fix: Tuist 설정 및 빌드 오류 해결을 위한 임시 수정

- Shared모듈에 flexlayout 추가(중복일 가능성 있음, 빌드 오류 해결)
- MainModule에 ExampleAppSetting 추가
- DomainOauth모듈의 Test에 빌드 오류 문제를 해결하기 위해 LoginRecordService 주입(Dummy 객체)
- FeelinConfirmButton의 폰트 적용방식 수정(빌드 오류 해결)

inwoodev committed:
add: 아티스트선택화면의 artistCell적용 여부 파악을 위해 FeatureModule의 example에 샘플 이미지 추가

inwoodev committed:
feat<FeelinArtistCell>: 아티스트 선택화면의 cell 구현

- 이미지 주변의 보더라인 설정을 외부에서 적용할 수 있는 구조로 구현

inwoodev committed:
feat<FeelinSearchBar>: 공용 검색 뷰 컴포넌트 제작

- figma에 명시된 padding, 폰트, 그리고 이미지를 활용하여 커스텀 검색뷰 구현

inwoodev committed:
add: FeatureMain의 example 모듈에 AppDelegate, SceneDelegate 추가

inwoodev committed:
feat<Artist> 아티스트 모델 구현 & 추가

inwoodev committed:
feat<ArtistSelectView> ArtistSelectView 구현 & 추가

inwoodev committed:
feat<ArtistSelectViewController> 아티스트 선택 ViewController 구현 & 추가

inwoodev commented on Projects/Feature/Main/Example/Resources/Assets.xcassets/AppIcon.appiconset/Contents.json:
```diff
@@ -0,0 +1,98 @@
+{
```
example app에서 아티스트 이미지를 볼 수 있도록 이미지파일들을 asset에 추가하였습니다.

inwoodev commented on Projects/Shared/DesignSystem/Sources/Cells/FeelinArtistCell.swift:
```diff
@@ -0,0 +1,101 @@
+        }
+    }
+    
+    public func configure(
+        artistName: String,
+        artistImage: UIImage?,
+        imageBorderWidth: CGFloat = 0,
+        imageBorderInset: CGFloat = 0
+    ) {
+        self.artistNameLabel.text = artistName
+        self.artistImageView.image = artistImage
+        self.borderWidth = imageBorderWidth
+        self.inset = imageBorderInset
+        self.setUpLayout()
+    }
+    
+    /// configure메서드를 활용하여 imageBorderWidth & imageBorderInset 설정이 되어있지 않으면 아래 메서드는 적용이 되지 않습니다.
+    public func setArtistBorder(color: UIColor) {
+        self.artistImageBorderView.layer.borderColor = color.cgColor
+    }
```
figma상 홈화면의 아티스트셀 같은 경우 아티스트 이미지에 별도의 border가 존재하지 않습니다.
따라서 상황에 따라 아티스트이미지 주변에 border를 주거나 주지 않을 수 있도록 cell을 구현하였습니다.

`setArtistBorder(color: UIColor)` 메서드를 활용하여 필요에 따라 해당 border색상을 변경할 수 있습니다.

derrickkim0109 replied:
해당 Cell을 홈에서 그대로 사용하면 되겠군요~! 굿입니다

inwoodev commented on Projects/Shared/DesignSystem/Sources/Views/FeelinSearchBar.swift:
```diff
@@ -0,0 +1,116 @@
+import FlexLayout
+import PinLayout
+
+final public class FeelinSearchBar: UIView {
```
figma에 있는 공용 SearchBar를 컴포넌트화 하여 구현하였습니다.

애플에서 기본으로 제공하는 SearchBar와 폰트, 이미지 그리고 textfield와 searchbar사이의 간격이 다르기에 커스텀으로 구현이 불가피하였습니다.

searchTextField 그리고 clearButton을 public하게 구현하여 외부 모듈에서도 접근할 수 있도록 구현하였습니다.

derrickkim0109 replied:
사소한 부분인데 final public의 위치가 각 객체별로 달라요..ㅎ

inwoodev replied:
@derrickkim0109 ㅋㅋㅋㅋ 어떻게 통일할까요~~?

derrickkim0109 replied:
final public, public final 둘 중 뭐든 상관없지만 한 가지 순서에 맞추면 좋지 않을까요~?

접근제어자가 먼저 오는건 어떠세요? public final이요.

inwoodev replied:
좋습니다~ 수정할게요!

inwoodev commented on Projects/Domain/Sources/Models/Artist.swift:
```diff
@@ -0,0 +1,25 @@
+//  Created by 황인우 on 6/16/24.
+//
+
+import UIKit
+
+public struct Artist: Hashable {
+    public var id: UUID = .init()
+    public var name: String
+    public var image: UIImage
+    public var isFavorite: Bool
+    
+    public init(
+        name: String,
+        image: UIImage,
+        isFavorite: Bool = false
+    ) {
+        self.name = name
+        self.image = image
+        self.isFavorite = false
+    }
+}
```
일단 cell에서 필요한 프로퍼티들로 artist모델을 구현하였습니다. 추후 수정하더라도 위의 프로퍼티들은 필요합니다.

ps: 추후 Domain의 하위모듈이 생길 경우 해당 모듈로 옮겨야 합니다.

inwoodev commented on Projects/Feature/Main/Interface/Sources/Views/ArtistSelectView.swift:
```diff
@@ -0,0 +1,133 @@
+        fatalError()
+    }
+    
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        
+        self.setUpArtistCollectionViewLayout()
+        
+        rootFlexContainer.pin
+            .top(self.pin.safeArea)
+            .horizontally()
+            .bottom()
+        rootFlexContainer.flex.layout()
+    }
```
해당 뷰의 프레임이 설정된 후 flowlayout의 크기를 지정할 수 있기에 layoutSubview메서드에 cell의 itemSize를 해당 메서드에서 지정합니다!

inwoodev commented on Projects/Feature/Main/Interface/Sources/ViewControllers/ArtistSelectViewController.swift:
```diff
@@ -0,0 +1,158 @@
+        self.artistCollectionView.delegate = self
+    }
+    
+    private func configureDataSource() {
+        let artistCellRegistration = UICollectionView.CellRegistration<FeelinArtistCell, Artist> { cell, indexPath, artist in
+            
+            cell.configure(
+                artistName: artist.name,
+                artistImage: artist.image,
+                imageBorderWidth: 2,
+                imageBorderInset: 5
+            )
+            cell.setArtistBorder(color: Colors.disabled)
+        }
+        
+        artistListDataSource = UICollectionViewDiffableDataSource(
+            collectionView: artistCollectionView,
+            cellProvider: { collectionView, indexPath, artist -> FeelinArtistCell in
+                return collectionView.dequeueConfiguredReusableCell(
+                    using: artistCellRegistration,
+                    for: indexPath,
+                    item: artist
+                )
+        })
+    }
```
diffableDatasource를 활용하여 컬렉션뷰의 datasource를 구성하였습니다. 그 이유는 해당 cell들은 추후 유저 검색에 따라 동적으로 갯수가 조정되어야 하기에 퍼포먼스 및 관리 측면에서 diffableDataSource가 더 수월하다 생각했기 때문입니다!

derrickkim0109 commented on Projects/Domain/Sources/Models/Artist.swift:
```diff
@@ -0,0 +1,25 @@
+import UIKit
+
+public struct Artist: Hashable {
+    public var id: UUID = .init()
```
id는 추후에 API의 id로 대체되는 걸까요?

inwoodev replied:
네네 그래야 할 것 같아요~!!

derrickkim0109 commented on Projects/Feature/Main/Example/Sources/SceneDelegate.swift:
```diff
@@ -0,0 +1,39 @@
+        
+        window = UIWindow(windowScene: windowScene)
+        
+//        window?.rootViewController = ArtistSelectViewController()
+        window?.makeKeyAndVisible()
```
주석 처리하신 이유가 있을까요?

inwoodev replied:
추후 MainModule에서 작업할 때 VC들이 많아질 경우 특정VC를 확인하고 싶은 경우 별도의 코드 작성 없이 간편하게 주석을 해제해서 볼 수 있도록 한 설정입니당:)

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/Views/ArtistSelectView.swift:
```diff
@@ -0,0 +1,133 @@
+
+import Shared
+
+class ArtistSelectView: UIView {
```
final 선언이 빠진것 같아요.

inwoodev replied:
오 수정할게요~!!

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/Views/ArtistSelectView.swift:
```diff
@@ -0,0 +1,133 @@
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError()
```
available(*, unavailable)이 빠져있어요~

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/ViewControllers/ArtistSelectViewController.swift:
```diff
@@ -0,0 +1,158 @@
+    
+    // MARK: - Diffable DataSource
+    
+    private enum ArtistListSection: CaseIterable {
```
ArtistListSection enum에 CaseIterable을 채택하신 이유가 있을까요? 
추후 제거될 예정인가요?

inwoodev replied:
Section자체를 순회할 수 있도록 CaseIterable타입을 채택하게 했어요. 추후 섹션이 추가되고 여러 section을 업데이트해야 할 경우 때 더 수월할 수 있겠다 판단을 했습니다 :)

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/ViewControllers/ArtistSelectViewController.swift:
```diff
@@ -0,0 +1,158 @@
+    
+    // MARK: - View
+    
+    var artistSelectView: ArtistSelectView = .init()
```
private 처리해도 되지 않을까요?

inwoodev replied:
맞습니다~

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/ViewControllers/ArtistSelectViewController.swift:
```diff
@@ -0,0 +1,158 @@
+        case main
+    }
+    
+    private var artistListDataSource: UICollectionViewDiffableDataSource<ArtistListSection, Artist>!
```
**P3**
- lazy var로 configureDataSource 메서드를 리턴 받는 방식은 어떠세요? 이 방식을 사용하면 강제 옵셔널 처리를 하지 않아도 될 것 같아서요

derrickkim0109 replied:
UICollectionViewDiffableDataSource<ArtistListSection, Artist>를 typealias 로 선언하고 configureDataSource 메서드에서 DataSource 타입을 리턴하는 방식을 말합니다~!

inwoodev replied:
로직을 한 군데로 모으기 적당한 방법 같아요~ 수정할게요!

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/ViewControllers/ArtistSelectViewController.swift:
```diff
@@ -0,0 +1,158 @@
+        
+        artists[indexPath.row] = updatedArtist
+        
+        updatedArtist.isFavorite ?
+        cell.setArtistBorder(color: Colors.active) :
+        cell.setArtistBorder(color: Colors.disabled)
```
삼항연산자는 앞으로 이 방식으로 통일하면 좋을 것 같네요~!ㅎ

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/Views/ArtistSelectView.swift:
```diff
@@ -0,0 +1,133 @@
+        return label
+    }()
+    
+    private let subtitlelabel: UILabel = {
```
이 프로퍼티는 카멜기법으로 네이밍이 작성되지 않았어요~

inwoodev replied:
수정해야겠네요!! :)

derrickkim0109 reviewed (commented):
너무 잘 만들어주신것 같아요. 수고하셨습니다~! 

자잘한 코멘트 남겼는데 확인 부탁드릴게요~!

noeyiz commented on Projects/Domain/Sources/Models/Artist.swift:
```diff
@@ -0,0 +1,25 @@
+    public var image: UIImage
+    public var isFavorite: Bool
+    
+    public init(
+        name: String,
+        image: UIImage,
+        isFavorite: Bool = false
+    ) {
+        self.name = name
+        self.image = image
+        self.isFavorite = false
+    }
```
혹시 요 부분 함수 내부에서도 false로 설정되는 게 맞나요 ..? :eyes:

inwoodev replied:
아니요`~ 이건 제 실수입니당 ㅎㅎ; 놓치고 나중에 엄청 삽질할 뻔했네요~ 짚어주셔서 감사해요 :)

noeyiz commented on Projects/Shared/DesignSystem/Sources/Cells/FeelinArtistCell.swift:
```diff
@@ -0,0 +1,101 @@
+
+import FlexLayout
+import PinLayout
+import SharedUtil
```
여기에서 SharedUtil 임포트가 필요한 이유가 있나용 ?

inwoodev replied:
없네요. 왜 포함되었는지 정말 모르겠네요 ㅋㅋㅋ. 수정할게요!!

noeyiz reviewed (commented):
- 아티스트셀을 재사용 가능하게 고려하여 구현하신 것
- 공용 SearchBar를 구현하신 것

확인했습니다 ! 👍🏻 

아주 사소한 질문과 함께 코멘트를 남겼습니당 수고하셨습니다 ~!

inwoodev committed:
refactor: Tuist 수정, 접근제어자 수정, 및 자잘한 수정

- TokenKeys.xcconfig를 메인 앱의 additionalFiles에 추가
- Artist객체 이니셜라이저 수정
- ArtistSelectViewController의 diffableDatasource 로직의 중앙화(lazy var 사용)
- 가독성 향상을 위한 typealist 활용
- 네이밍 수정(카멜케이스 적용)
- 불필요한 import제거
- 몇 몇 접근제어자들을 적절한 범위로 제한

inwoodev commented:
@noeyiz  @derrickkim0109 
여러분의 소중한 코멘트 꼼꼼히 읽어보고 다음과 같은 수정사항을 적용했습니다 :)
- TokenKeys.xcconfig를 메인 앱의 additionalFiles에 추가
- Artist객체 이니셜라이저 수정
- ArtistSelectViewController의 diffableDatasource 로직의 중앙화(lazy var 사용)
- 가독성 향상을 위한 typealist 활용
- 네이밍 수정(카멜케이스 적용)
- 불필요한 import제거
- 몇 몇 접근제어자들을 적절한 범위로 제한

추가적인 수정사항이 필요하다면 코멘트나 슬랙 주세요~. 괜찮다 판단 되시면 approve 부탁드립니다~!

noeyiz reviewed (approved):
확인했습니다 ~ 고생하셨습니다 ! 👍🏻
